src/analysis/complex/isometry.lean

```diff
@@ -0,0 +1,171 @@
+
+## References
+
+* [Isometries of the Complex Plane](http://helmut.knaust.info/mediawiki/images/b/b5/Iso.pdf)
+-/
+
+open complex
+
+local notation `|` x `|` := complex.abs x
+
+lemma add_self_eq (a : ℝ) : a + a = a * 2 := by ring
+
+lemma linear_isometry_complex_aux (z : ℂ) (f : ℂ →ₗᵢ[ℝ] ℂ) :
+  f 0 = 0 → f 1 = 1 → ((f z = z) ∨ (f z = conj z)) :=
+begin
+  intros h₀ h₁,
+  have hf0 : |f z - f 0| = |z - 0| := by exact linear_isometry.dist_map f z 0,
+  rw h₀ at hf0,
+  iterate 2 { rw sub_zero at hf0 },
+  have hf1 : |f z - f 1| = |z - 1| := by exact linear_isometry.dist_map f z 1,
+  rw h₁ at hf1,
+  have h₂ : conj (f z - 1) * (f z - 1) = conj (z - 1) * (z - 1) :=
+    begin
+      iterate 2 { rw ← norm_sq_eq_conj_mul_self },
+      iterate 2 { rw ← mul_self_abs },
+      rw hf1,
+    end,
+  iterate 2 { rw conj_sub at h₂ },
+  iterate 2 { rw sub_mul at h₂ },
+  iterate 4 { rw mul_sub at h₂ },
+  rw conj_one at h₂,
+  rw mul_one at h₂,
+  iterate 2 { rw one_mul at h₂ },
+  rw mul_one at h₂,
+  rw one_mul at h₂,
+  iterate 2 { rw ← norm_sq_eq_conj_mul_self at h₂ },
+  iterate 2 { rw ← mul_self_abs at h₂ },
+  iterate 2 { rw ← sub_add at h₂ },
+  rw hf0 at h₂,
+  have h₃ : z + conj z = f z + conj (f z) := by {
+    iterate 2 { rw ← sub_add_eq_sub_sub at h₂ },
+    iterate 2 { rw sub_add at h₂ },
+    rw ← neg_add_eq_sub at h₂,
+    rw ← neg_add_eq_sub (conj z + z - 1) at h₂,
+    rw add_right_cancel_iff at h₂,
+    iterate 2 { rw add_sub_assoc at h₂ },
+    iterate 2 { rw neg_add' at h₂ },
+    iterate 2 { rw ← sub_add at h₂ },
+    rw add_right_cancel_iff at h₂,
+    iterate 2 { rw ← neg_add' at h₂ },
+    rw add_comm at h₂,
+    rw add_comm (conj z) z at h₂,
+    rw ← @add_left_cancel_iff _ _ (z + conj z) at h₂,
+    rw add_neg_self at h₂,
+    rw ← @add_right_cancel_iff _ _ (f z + conj (f z)) at h₂,
+    rw zero_add at h₂,
+    rw add_assoc at h₂,
+    rw neg_add_self at h₂,
+    rw add_zero at h₂,
+    rw h₂,
+  },
+  have hf_re : (f z).re = z.re := by {
+    rw ext_iff at h₃,
+    iterate 2 { rw add_re at h₃ },
+    iterate 2 { rw add_im at h₃ },
+    iterate 2 { rw conj_re at h₃ },
+    iterate 2 { rw conj_im at h₃ },
+    iterate 2 { rw add_neg_self at h₃ },
+    iterate 2 { rw add_self_eq at h₃ },
+    cases h₃,
+    rw eq_comm at h₃_left,
+    linarith [h₃_left],
+  },
+  have hf_im : (f z).im = z.im ∨ (f z).im = -z.im := by {
+    iterate 2 { rw complex.abs at hf0 },
+    rw real.sqrt_inj at hf0,
+    swap,
+    exact norm_sq_nonneg (f z),
+    swap,
+    exact norm_sq_nonneg z,
+    rw norm_sq_apply (f z) at hf0,
+    rw norm_sq_apply z at hf0,
+    rw hf_re at hf0,
+    rw add_left_cancel_iff at hf0,
+    rw mul_self_eq_mul_self_iff at hf0,
+    exact hf0,
+  },
+  iterate 2 { rw ext_iff },
+  rw conj_re,
+  rw conj_im,
+  rw ← and_or_distrib_left,
+  exact ⟨hf_re, hf_im⟩,
+end
+
+lemma linear_isometry_complex (z : ℂ) (f : ℂ →ₗᵢ[ℝ] ℂ) :
+  ∃ a : ℂ, |a| = 1 ∧ ((f z = a * z) ∨ (f z = a * conj z)) :=
```

Comment:
This statement is not as strong as it should be -- can you adjust it?  You've stated a theorem in which the `a` depends on the `z`, but in fact the `a` should be independent of `z`:
```lean
lemma linear_isometry_complex (f : ℂ →ₗᵢ[ℝ] ℂ) :
  ∃ a : ℂ, |a| = 1 ∧ ((∀ z, f z = a * z) ∨ (∀ z, f z = a * conj z)) :=
```

Reply:
Now I think of it, it might be useful in any case to give a separate definition of the rotation linear-isometries, and this allows for a more elegant statement (but equivalent to the other one I said):
```lean
def rotation {a : ℂ} (ha : |a| = 1) : ℂ →ₗᵢ[ℝ] ℂ :=
{ to_fun := λ z, a * z,
  etc }

lemma linear_isometry_complex (f : ℂ →ₗᵢ[ℝ] ℂ) :
  ∃ a : ℂ, |a| = 1 ∧ (f = rotation a ∨ f = (rotation a).comp isometry_conj) :=
```

Reply:
> This statement is not as strong as it should be -- can you adjust it? You've stated a theorem in which the `a` depends on the `z`, but in fact the `a` should be independent of `z`:
> 
> ```lean
> lemma linear_isometry_complex (f : ℂ →ₗᵢ[ℝ] ℂ) :
>   ∃ a : ℂ, |a| = 1 ∧ ((∀ z, f z = a * z) ∨ (∀ z, f z = a * conj z)) :=
> ```

I made the quoted change, however I haven't yet made [this change to using `rotation`](https://github.com/leanprover-community/mathlib/pull/6923#discussion_r602892917). Can I get what I have now reviewed before moving on to the next change? Or is it preferable to wait for the change to using `rotation`? Thanks!